tests/python_package_test/test_dask.py

```diff
@@ -656,6 +658,169 @@ def test_ranker(output, client, group):
+            if 'dataframe' in output:
+                d_init_score = dy_e.map_partitions(lambda x: pd.Series([init_score_value] * x.size))
+            else:
+                d_init_score = dy_e.map_blocks(lambda x: np.repeat(init_score_value, x.size))
+
+            eval_init_score.append(d_init_score)
+
+    fit_trees = 50
+    params = {
+        "random_state": 42,
+        "n_estimators": fit_trees,
+        "num_leaves": 2
+    }
+
+    model_factory = task_to_dask_factory[task]
+    dask_model = model_factory(
+        client=client,
+        **params
+    )
+
+    fit_params = {
+        'X': dX,
+        'y': dy,
+        'eval_set': eval_set,
+        'eval_names': eval_names,
+        'eval_sample_weight': eval_sample_weight,
+        'eval_init_score': eval_init_score,
+        'eval_metric': eval_metrics,
+        'verbose': True
+    }
+    if task == 'ranking':
+        fit_params.update(
+            {'group': dg,
+             'eval_group': eval_group,
+             'eval_at': eval_at}
+        )
+    elif task == 'binary-classification':
+        fit_params.update({'eval_class_weight': eval_class_weight})
+
+    if eval_sizes == [0]:
+        with pytest.warns(UserWarning, match='Worker (.*) was not allocated eval_set data. Therefore evals_result_ and best_score_ data may be unreliable.'):
+            dask_model.fit(**fit_params)
+    else:
+        dask_model = dask_model.fit(**fit_params)
+
+        # total number of trees scales up for ova classifier.
+        if task == 'multiclass-classification':
+            model_trees = fit_trees * dask_model.n_classes_
+        else:
+            model_trees = fit_trees
+
+        # check that early stopping was not applied.
+        assert dask_model.booster_.num_trees() == model_trees
+        assert dask_model.best_iteration_ is None
+
+        # checks that evals_result_ and best_score_ contain expected data and eval_set names.
+        evals_result = dask_model.evals_result_
+        best_scores = dask_model.best_score_
+        assert len(evals_result) == n_eval_sets
+        assert len(best_scores) == n_eval_sets
+
+        for eval_name in evals_result:
+            assert eval_name in dask_model.best_score_
+            if eval_names:
+                assert eval_name in eval_names + ['valid_2']
+            else:
+                if eval_name == 'training':
+                    assert 1 in eval_sizes
+                else:
+                    eval_name.startswith('valid')
+
+            # check that each of eval_metrics calculated on all eval sets.
+            for i, metric in enumerate(eval_metrics):
+                if task == 'ranking':
+                    metric += f'@{eval_at[i]}'
+
+                assert metric in evals_result[eval_name]
+                assert metric in best_scores[eval_name]
+                assert len(evals_result[eval_name][metric]) == fit_trees
+
+    client.close(timeout=CLIENT_CLOSE_TIMEOUT)
+
```

Comment:
While testing this, I found something interesting that I think could be a source of instability in these tests and in real-world use of this feature.

It seems like `evals_result_` will only contain the evaluation results for the pieces of the evaluation data that happened to end up on the worker that is `rank 0`. If no pieces of the eval sets ended up on that worker, `evals_result_` will be `None`.

See the links below for how that worker is treated slightly differently:

* https://github.com/microsoft/LightGBM/blob/e4cf2e4feb2f2cb29cec55070febd74ba8fa28cc/python-package/lightgbm/dask.py#L325
* https://github.com/microsoft/LightGBM/blob/e4cf2e4feb2f2cb29cec55070febd74ba8fa28cc/python-package/lightgbm/dask.py#L400
* https://github.com/microsoft/LightGBM/blob/e4cf2e4feb2f2cb29cec55070febd74ba8fa28cc/python-package/lightgbm/dask.py#L123

This is a tough situation to replicate, but I think you have a good chance of replicating it by using an evaluation set with a single partition and many workers.  If you try running this a few times in a row, one of them should demonstrate this situation.

```python
import dask.array as da
import lightgbm as lgb
import numpy as np
from sklearn.datasets import make_regression
from distributed import Client, LocalCluster, wait

cluster = LocalCluster(n_workers=5)
client = Client(cluster)

# make sure you're starting with clean workers every time
client.restart()

# tightly control sizes of chunks, just for reproducibility purposes
X_np, y_np = make_regression(n_samples=1000, n_features=10)
X = da.from_array(X_np, chunks=(100, 10))
y = da.from_array(y_np, chunks=(100,))

# create an evaluation set with a single partitions
X_e = da.random.random((100, 10))
y_e = da.random.random((100,))

# prove to ourselves that this evaluation set has a single partitions
print(X_e.npartitions)

model = lgb.DaskLGBMRegressor(
    num_iterations=5,
)

model.fit(
    X,
    y,
    eval_metric=["rmse"],
    eval_set=[
        (X_e, y_e)
    ]
)

str(model.evals_result_)
```

![image](https://user-images.githubusercontent.com/7608904/112927295-5e0b4800-90da-11eb-9f3e-c8f0e7feb19f.png)


<hr>

This is made worse by the fact that `lightgbm.dask` doesn't give you a way to know which worker is going to end up being rank 0 (intentionally). So you, as a user, couldn't even say "I know about this limitation, I'm going to just make sure my eval data is all on the rank 0 worker" or something.

I'm not sure what we should do about this, I need to think about it a bit. I can think of a few options, I'm just listing them here to maybe inspire creativity, without pre-filtering ideas that I think are bad.

1. edit https://github.com/microsoft/LightGBM/blob/e4cf2e4feb2f2cb29cec55070febd74ba8fa28cc/python-package/lightgbm/dask.py#L123 to return the model from every `_train_part()` task, and store the "non-rank-0" `evals_result_` data on a new property in the Dask model objects like `.other_evals_` or something like that. So accessing `.evals_result_` will look and feel just like the sklearn API, but you can go find the other evaluation results in that other property if you want them.
2. try to write a different evaluation callback for Dask specifically, which, after each boosting round, aggregates evaluation metrics across all workers somehow
3. when training is over, before exiting `fit()` combine all the evaluation results by taking something like a mean weighted by sample count. This probably would give misleading or inaccurate results with some metrics though :/.
4. change nothing and just say "this is how it works, copy your eval data onto every Dask worker if you want to be sure you get eval results"

Reply:
Yes, I agree that a worker that received no `eval_set` data starts causing lots of issues, the `eval_result_` issue just one among them. This goes to the test case with the `pytest.warns`, when `eval_sizes = [0]` and no worker gets _any_ eval data.

I'm confused how you managed to end up with a worker with "no pieces of the eval sets ended up on that worker"? It should be the case that if no pieces of the eval set end up on rank 0, this would cause the test to fail because `dask.py` throws the error when it finds out that any worker is allocated 0 partitions of eval_set data.

The rank 0 thing raises the issue that perhaps one sub-dataset of an eval_set whose length is > 1 might not receive any parts of a particular sub-dataset, in which case `len(evals_result)` could be `< len(eval_set)`, for example, if there are two `(X, y)` pairs in `eval_set` but the rank 0 worker only gets parts allocated to it from a single pair. This `None` case should be eliminated, but I agree that the randomness of what eval_results_ set you get is undesirable.

Reply:
Oh I'm so sorry, I just realized you already had the text "Therefore evals_result_ and best_score_ data may be unreliable" in the warning message! Think I just glazed over that when I read it and thought "ok yeah yeah this says something about not getting eval data, cool" and didn't read it carefully.

I can help look at `xgboost.dask` as well and try to help. We have a good relationship with the XGBoost maintainers and can also `@` them in here if we need another opinion.

Reply:
On the subject of nondeterministic returning of `evals_set_` - it looks like `xgboost.dask` does something similar to the current `lightgbm.dask` implementation, but differs in two main aspects:

1. Each worker returns `booster` and `history` (xgboost's `evals_result_` equivalent) in a 2-element dict
2. Filtering of returned workers' estimators is not dependent on a rank number:
```return list(filter(lambda ret: ret is not None, results))[0]```

I think "eval_set training artifacts handling" could be a separate PR though, don't see it as a blocker for this PR

Reply:
I agree, I think it's desirable to handle in a follow-up PR. This PR is adding the main infrastructure for a fairly large chunk of functionality, so I agree with the idea of merging it once it's mostly there and handling things like this in followups.